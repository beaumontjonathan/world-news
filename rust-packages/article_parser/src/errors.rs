use chrono::ParseError as ChronoParseError;
use htmlescape::DecodeErr;
use scraper::Selector;
use serde_json::Error as SerdeJsonError;

#[derive(Debug)]
pub enum InternalParseError {
  UnknownError(&'static str),
  HtmlStructureError(HtmlStructureError),
  JsonParseError(JsonParseError),
  InvalidDateTime(ChronoParseError),
  HtmlDecodeErr(DecodeErr),
}

#[derive(Debug)]
pub enum JsonParseErrorCause {
  InvalidLdJson,
  InvalidDataAttrJson,
}

#[derive(Debug)]
pub struct JsonParseError {
  pub cause: JsonParseErrorCause,
  pub serde_error: SerdeJsonError,
}

#[derive(Debug)]
pub enum HtmlErrorCause {
  MissingElement,
  NonUniqueElement,
  MissingAttribute(String),
}

#[derive(Debug)]
pub struct HtmlStructureError {
  pub cause: HtmlErrorCause,
  pub selector: &'static Selector,
}

impl InternalParseError {
  pub fn html(error: HtmlStructureError) -> InternalParseError {
    InternalParseError::HtmlStructureError(error)
  }

  pub fn json(error: JsonParseError) -> InternalParseError {
    InternalParseError::JsonParseError(error)
  }

  pub fn ld_json(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidLdJson,
      serde_error,
    })
  }

  pub fn json_data_attr(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidDataAttrJson,
      serde_error,
    })
  }

  pub fn unknown(s: &'static str) -> InternalParseError {
    InternalParseError::UnknownError(s)
  }

  pub fn chrono(err: ChronoParseError) -> InternalParseError {
    InternalParseError::InvalidDateTime(err)
  }

  pub fn decode_html(err: DecodeErr) -> InternalParseError {
    InternalParseError::HtmlDecodeErr(err)
  }
}

impl From<ChronoParseError> for InternalParseError {
  fn from(err: ChronoParseError) -> Self {
    InternalParseError::chrono(err)
  }
}

impl From<DecodeErr> for InternalParseError {
  fn from(err: DecodeErr) -> Self {
    InternalParseError::decode_html(err)
  }
}
