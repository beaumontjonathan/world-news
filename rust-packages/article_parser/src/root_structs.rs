use chrono::{DateTime, Utc};
use serde::Serialize;
use std::fmt::Debug;

#[derive(Debug, Serialize)]
pub struct Image {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct Video {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
pub enum Publisher {
  BBC,
  DailyMail,
  Independent,
  Metro,
  Mirror,
  Sky,
  Sun,
  Guardian,
}

use Publisher::*;

impl From<String> for Publisher {
  fn from(publisher: String) -> Self {
    match publisher.to_lowercase().as_str() {
      "bbc" => BBC,
      "dailymail" => DailyMail,
      "independent" => Independent,
      "metro" => Metro,
      "mirror" => Mirror,
      "sky" => Sky,
      "sun" => Sun,
      "theguardian" | "guardian" => Guardian,
      _ => panic!("Invalid publisher type"),
    }
  }
}

#[derive(Debug, Serialize)]
pub struct NewsArticleContent {
  pub headline: String,
  pub twitter_headline: Option<String>,
  pub description: String,
  pub thumbnail: Image,
  pub categories: Vec<String>,
  pub images: Vec<Image>,
  pub videos: Vec<Video>,
  pub body: String,
  pub date_updated: DateTime<Utc>,
  pub date_published: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct VideoArticleContent {
  pub something: String,
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, Serialize)]
pub enum ParsedPageContent {
  NewsArticle(NewsArticleContent),
  VideoArticle(VideoArticleContent),
}

#[derive(Debug, Serialize)]
pub struct PageMeta {
  pub publisher: Publisher,
  pub url: String,
  pub date_parsed: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct ParsedPage {
  pub meta: PageMeta,
  pub content: ParsedPageContent,
}
