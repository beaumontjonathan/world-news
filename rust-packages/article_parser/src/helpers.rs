use crate::errors::{HtmlErrorCause, HtmlStructureError, InternalParseError};
// use chrono::{DateTime, Utc};
use crate::root_structs::ParsedPageContent;
use lazy_static::lazy_static;
use regex::Regex;
use scraper::{ElementRef, Html, Selector};
// use std::fmt::{self, Debug, Display, Formatter};
pub trait ToOwnedOrNone {
  fn to_owned_or_none(self) -> Option<String>;
}

impl<'a> ToOwnedOrNone for Option<&'a str> {
  fn to_owned_or_none(self) -> Option<String> {
    self.and_then(|mut s| {
      s = s.trim();
      if s.is_empty() {
        None
      } else {
        Some(String::from(s))
      }
    })
  }
}
// pub trait NoneIfEmpty
// where
//   Self: Sized,
// {
//   fn none_if<F>(self, callback: F) -> Option<Self>
//   where
//     F: Fn(&Self) -> bool,
//   {
//     if callback(&self) {
//       None
//     } else {
//       Some(self)
//     }
//   }
// }
// impl<T> NoneIfEmpty for T {}

pub type ParseResult = Result<ParsedPageContent, InternalParseError>;

pub trait FixHtmlWhitespace {
  fn fix_html_whitespace(self) -> Self;
}

fn fix_whitespace(text: String) -> String {
  lazy_static! {
    static ref RE: Regex = Regex::new(r#"\s+"#).unwrap();
  }
  RE.replace_all(&text, " ").into_owned()
}

impl FixHtmlWhitespace for String {
  fn fix_html_whitespace(self) -> String {
    fix_whitespace(self)
  }
}

pub trait AttrValueOrNone {
  fn attr_value_or_none(&self, attr: &str) -> Option<String>;
}

impl<'a> AttrValueOrNone for ElementRef<'a> {
  fn attr_value_or_none(&self, attr: &str) -> Option<String> {
    self.value().attr(attr).to_owned_or_none()
  }
}

pub trait FullFixedHtmlText {
  fn full_fixed_html_text(self) -> String;
}

impl<'a> FullFixedHtmlText for ElementRef<'a> {
  fn full_fixed_html_text(self) -> String {
    self
      .text()
      .map(|text| text.trim().to_string())
      .collect::<Vec<String>>()
      .join(" ")
      .fix_html_whitespace()
      .trim()
      .to_owned()
  }
}

pub trait JoinWithNewline {
  fn join_with_newline(self) -> String;
}

impl<T> JoinWithNewline for T
where
  T: Iterator<Item = String>,
{
  fn join_with_newline(self) -> String {
    self.collect::<Vec<String>>().join("\n").trim().to_owned()
  }
}

pub trait UniqueElementRef<'a> {
  fn unique(self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause>;
}

impl<'a, T> UniqueElementRef<'a> for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn unique(mut self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause> {
    match (self.next(), self.next()) {
      (None, _) => Err(HtmlErrorCause::MissingElement),
      (_, Some(_)) => Err(HtmlErrorCause::NonUniqueElement),
      (Some(el_ref), None) => Ok(el_ref),
    }
  }
}

pub trait SelectHelpers {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError>;
}

impl SelectHelpers for Html {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

impl SelectHelpers for scraper::element_ref::ElementRef<'_> {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

pub trait FormattedText {
  fn formatted_text(self) -> String;
}

impl<'a, T> FormattedText for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn formatted_text(self) -> String {
    self.map(|p| p.full_fixed_html_text()).join_with_newline()
  }
}

pub trait ChildrenHelpers {
  fn children_first(&self) -> Option<ElementRef>;
  fn children_all(&self) -> Vec<ElementRef>;
}

impl ChildrenHelpers for scraper::element_ref::ElementRef<'_> {
  fn children_first(&self) -> Option<ElementRef> {
    self.children_all().into_iter().next()
  }

  fn children_all(&self) -> Vec<ElementRef> {
    self.children().filter_map(ElementRef::wrap).collect()
  }
}
