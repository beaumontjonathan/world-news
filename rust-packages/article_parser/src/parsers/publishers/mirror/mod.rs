use crate::errors::InternalParseError;
use crate::helpers::{
  AttrValueOrNone, FullFixedHtmlText, JoinWithNewline, ParseResult, SelectHelpers,
};
use crate::my_selectors;
use crate::root_structs::{Image, NewsArticleContent, ParsedPageContent, Video};
use chrono::DateTime;
use scraper::{Html, Selector};

my_selectors! {
  ARTICLE_ROOT_SELECTOR => "body > main > article.article-main",
  ARTICLE_CONTENT_SELECTOR => ":scope > div.article-wrapper > div.content-column",
  ARTICLE_BODY_SELECTOR => ":scope > div.article-body",
  ARTICLE_CONTENT_CHILDREN_SELECTOR => ":scope > *",
  H1_SELECTOR => "h1",
  SUB_TITLE_SELECTOR => "p.sub-title",
  THUMBNAIL_FIGURE_SELECTOR => "figure.lead-article-image",
  IMG_SELECTOR => "img",
  CAPTION_SELECTOR => "figcaption > .caption",
  URL_SELECTOR => "meta[itemprop=\"url\"]",
  ARTICLE_BODY_FIGURE_IMAGE_SELECTOR => ".article-body > figure.in-article-image",
  ARTICLE_BODY_P_AND_URL_SELECTOR => ":scope > p, :scope > ul",
  DATE_UPDATED_SELECTOR => "head > meta[property=\"article:modified_time\"]",
  DATE_PUBLISHED_SELECTOR => "head > meta[property=\"article:published_time\"]",
  DIV_MOD_VIDEO => "div.mod-video",
  FIGURE_LEAD_ARTICLE_IMAGE => "figure.lead-article-image",
  FIGURE_IN_ARTICLE_IMAGE => ":scope > figure.in-article-image",
}

pub fn parse(document: Html) -> ParseResult {
  // DOCUMENT SELECTED
  let article_root = document.select_unique(&ARTICLE_ROOT_SELECTOR)?;
  let article_content = article_root.select_unique(&ARTICLE_CONTENT_SELECTOR)?;
  let article_body = article_content.select_unique(&ARTICLE_BODY_SELECTOR)?;

  // HEADLINE
  let headline = article_root
    .select_unique(&H1_SELECTOR)?
    .full_fixed_html_text();

  // TWITTER HEADLINE
  let twitter_headline = None;

  // DESCRIPTION
  let description: String = article_root
    .select_unique(&SUB_TITLE_SELECTOR)
    .map(|sub_title| sub_title.full_fixed_html_text())
    .unwrap_or_else(|_| String::new());

  // THUMBNAIL
  let thumbnail: Image = {
    let first_child = article_content.select_first(&ARTICLE_CONTENT_CHILDREN_SELECTOR)?;
    if DIV_MOD_VIDEO.matches(&first_child) {
      // todo - could get data from the "div.json-placeholder" element, but
      // html escaping breaks the json - e.g., double quotes in headline text.
      Ok(Image {
        alt: None,
        url: None,
        caption: None,
      })
    } else if FIGURE_LEAD_ARTICLE_IMAGE.matches(&first_child) {
      let caption = Some(
        first_child
          .select_unique(&CAPTION_SELECTOR)?
          .full_fixed_html_text(),
      );
      let img = first_child.select_unique(&IMG_SELECTOR)?;
      let url = img.attr_value_or_none("src");
      let alt = img.attr_value_or_none("alt");
      Ok(Image { alt, url, caption })
    } else {
      Err(InternalParseError::unknown("thumbnail type unknown"))
    }
  }?;

  // CATEGORIES
  // TODO - implement
  let categories: Vec<String> = vec![];

  // IMAGES
  let images = article_body
    .select(&FIGURE_IN_ARTICLE_IMAGE)
    .map(|figure| {
      let img = figure.select_first(&IMG_SELECTOR)?;
      let alt = img.value().attr("alt").and_then(|s| match s.len() {
        0 => None,
        _ => Some(s.to_owned()),
      });
      let url = figure
        .select_unique(&URL_SELECTOR)?
        .attr_value_or_none("content");
      let caption = figure
        .select(&CAPTION_SELECTOR)
        .next()
        .map(|el| el.full_fixed_html_text());
      Ok(Image { url, alt, caption })
    })
    .collect::<Result<_, InternalParseError>>()?;

  // VIDEOS
  // todo - implement
  let videos: Vec<Video> = vec![];

  // BODY
  let body = article_body
    .select(&ARTICLE_BODY_P_AND_URL_SELECTOR)
    .map(|el| el.full_fixed_html_text())
    .join_with_newline();

  // DATE UPDATED
  let date_updated = document.select_unique_attr(&DATE_UPDATED_SELECTOR, "content")?;
  let date_updated = DateTime::parse_from_rfc3339(&date_updated)?.into();

  //DATE PUBLISHED
  let date_published = document.select_unique_attr(&DATE_PUBLISHED_SELECTOR, "content")?;
  let date_published = DateTime::parse_from_rfc3339(&date_published)?.into();

  Ok(ParsedPageContent::NewsArticle(NewsArticleContent {
    headline,
    twitter_headline,
    description,
    thumbnail,
    categories,
    images,
    videos,
    body,
    date_updated,
    date_published,
  }))
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::test_helpers::helpers::read_html_doc;
  use insta::assert_debug_snapshot;

  #[test]
  fn test_file_01() {
    let document = read_html_doc("test-01.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_02() {
    let document = read_html_doc("test-02.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_03() {
    let document = read_html_doc("test-03.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_04() {
    let document = read_html_doc("test-04.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_06() {
    let document = read_html_doc("test-06.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_07() {
    let document = read_html_doc("test-07.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }
}
