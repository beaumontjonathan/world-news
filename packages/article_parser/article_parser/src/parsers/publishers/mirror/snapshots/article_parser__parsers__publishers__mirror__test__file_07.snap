---
source: src/parsers/publishers/mirror/mod.rs
expression: article
---
Ok(
    NewsArticle(
        NewsArticleContent {
            headline: "BBC election debate: 6 key moments from the final head-to-head clash",
            twitter_headline: None,
            description: "Boris Johnson and Jeremy Corbyn clashed in their final head-to head before Thursday\'s poll. The two leaders faced each other in a Question Time style debate chaired by Nick Robinson",
            thumbnail: Image {
                alt: None,
                url: None,
                caption: None,
            },
            categories: [],
            images: [
                Image {
                    alt: None,
                    url: Some(
                        "https://i2-prod.mirror.co.uk/incoming/article21043028.ece/ALTERNATES/s615b/0_General-Election-2019.jpg",
                    ),
                    caption: Some(
                        "A member of the audience shouted \"go on then\"",
                    ),
                },
                Image {
                    alt: None,
                    url: Some(
                        "https://i2-prod.mirror.co.uk/incoming/article21042992.ece/ALTERNATES/s615b/0_Britains-PM-Johnson-and-Labour-leader-Corbyn-debate-on-BBC-in-London.jpg",
                    ),
                    caption: None,
                },
                Image {
                    alt: None,
                    url: Some(
                        "https://i2-prod.mirror.co.uk/incoming/article21042141.ece/ALTERNATES/s615b/0_BBC.jpg",
                    ),
                    caption: Some(
                        "Jeremy Corbyn says the NHS is not safe in the hands of the Tories",
                    ),
                },
                Image {
                    alt: None,
                    url: Some(
                        "https://i2-prod.mirror.co.uk/incoming/article21042980.ece/ALTERNATES/s615b/0_Britains-PM-Johnson-and-Labour-leader-Corbyn-debate-on-BBC-in-London.jpg",
                    ),
                    caption: Some(
                        "Jeremy Corbyn shed some light on Labour\'s much touted four-day week policy",
                    ),
                },
                Image {
                    alt: None,
                    url: Some(
                        "https://i2-prod.mirror.co.uk/incoming/article21042992.ece/ALTERNATES/s615b/0_Britains-PM-Johnson-and-Labour-leader-Corbyn-debate-on-BBC-in-London.jpg",
                    ),
                    caption: None,
                },
                Image {
                    alt: None,
                    url: Some(
                        "https://i2-prod.mirror.co.uk/incoming/article21012548.ece/ALTERNATES/s615b/16711581_10158181129255425_230962053886286278_n.jpg",
                    ),
                    caption: Some(
                        "Jack Merritt was stabbed to death during the London Bridge terror attack",
                    ),
                },
                Image {
                    alt: None,
                    url: Some(
                        "https://i2-prod.mirror.co.uk/incoming/article21042159.ece/ALTERNATES/s615b/1_THP_CHP_061219Slug_2960JPG.jpg",
                    ),
                    caption: Some(
                        "Boris Johnson went off on a strange tangent",
                    ),
                },
            ],
            videos: [],
            body: "Boris Johnson and Jeremy Corbyn clashed over Brexit , the NHS and terrorism in the final TV debate of the General Election campaign.\nThe Prime Minister accused Mr Corbyn of a \"failure of leadership\" in his handling of anti-Semitism in the Labour Party.\nThe Labour leader shot back that he never used racist language and that he hoped Mr Johnson regretted his past comments.\nViewers of the BBC TV debate believe the Prime Minister narrowly won although it was within the polling margin of error so is effectively too close to call.\nAccording to a YouGov snap poll 52% believe Mr Johnson won and 48% think Mr Corbyn did.\nThe pollster said this falls within the margin of error so it can be considered another draw. Those who answered \"don\'t know\" have been removed from results.\nChris Curtis, YouGov\'s political research manager, said: \"Our snap poll shows that the public remains divided on who won the debate, just as with last month\'s head-to-head, with most Labour voters thinking Jeremy Corbyn won, most Conservative voters thinking Boris Johnson won, and very few people changing their minds.\n\"But given the Conservatives went into this debate in the lead, they will hope the lack of a knockout blow means they can maintain this until voting day.\"\nBoris Johnson had a toe-curling response when asked what the punishment should be for politicians who lie.\nThe Prime Minister has been accused of misleading the country on a number of issues including his promises to create 50,000 nurse places - 20,000 of whom are already working in the service, his promise for 40 hospitals which transpired as only six and the infamous message on the side of the Vote Leave bus.\nAt the BBC\'s leaders\' debate in Maidstone, Kent, a member of the audience asked the two leaders what punishment they think is \"appropriate for elected politicians who lie during political campaigns\"?\nThe audience laughed at Mr Johnson as he smirked and said politicians who lie during election campaigns should \"be made to go on their knees down on their knees through the chamber of the House of Commons scourging themselves with copies of their offending documents which claim to prove one thing and actually prove something quite different.\"\nOne audience member shouted out: \"Go on then.\"\nJeremy Corbyn hit back at Boris Johnson\'s claims that the NHS is not on the table in future trade deals with the US.\nDuring the leadership debate the Labour leader was challenged on his assertion that if there was a deal with America then Brits would face five-figure bills for cancer treatment.\nMr Corbyn asked the Prime Minister why negotiations had taken two years if Mr Johnson was telling the US that the NHS was not for sale.\nHe said: \"The Prime Minister says he\'s not going to do that kind of trade deal. \"If that\'s the case, why did the talks go on for two years? It doesn\'t take two years to say no to privatisation of the NHS.\"\nThe Labour leader clashed with the Prime Minister during the BBC\'s head-to-head debate less than a week before the election.\nHost Nick Robinson asked: You don\'t dispute what he\'s quoted which is that American trade negotiators say they want more access to the NHS.\nHe replied: \"I\'ve made it absolutely clear to everybody that under no circumstances will that happen.\"\nMr Robinson asked Mr Corbyn: \"You sent an email on behalf of the Labour Party saying \'imagine opening a five figure bill for your cancer treatment\', do you accept that is totally misleading - there\'s no possibility of anybody having NHS cancer treatments and having to pay a five figure bill\".\nMr Corbyn replied: \"If you end up with a privatised service, that\'s what you get. If you end up with a US trade deal, that\'s what you get.\nMr Johnson said: \"This is pure Bermuda Triangle stuff. We\'ve heard it time and time again from the Labour Party during this election campaign.\n\"We\'ll be hearing about \'little green men\' next.\"\nMr Johnson added: \"Under no circumstances will we sell it off to anybody in any kind of trade deal.\"\nThe PM added that he has \"made it absolutely clear to everybody\" that under no circumstances will it happen.\nBoris Johnson was once again skewered on his bogus pledge to have 40 new hospitals (when in fact most of it is for seed funding).\nBut when a nurse asked about dealing with the shortage of staff, it was actually Jeremy Corbyn who shed some new light on policy.\nUnder pressure from BBC host Nick Robinson, Mr Corbyn confirmed Labour\'s ambition to reach a 32-hour week within a decade was not a hard plan to apply to NHS staff.\nMr Corbyn said he wants to see the working week come down, but added: \"There is no plan to bring in a four day week in the NHS.\"\nThis appears to contradict Shadow Chancellor John McDonnell who said last month: \"It’s a 32-hour week implemented over a ten-year period – it will apply to everyone.\"\nJeremy Corbyn found himself on stronger ground when talking about capitalism and its failings.\nAfter Boris Johnson accused him and his allies of wanting to overthrow capitalism he replied: “My plan is one where you don’t leave people behind.\n“You don’t ignore the fact that 4million of our children are living in poverty, you don’t ignore the fact there are thousands of people sleeping on the streets of this country - the fifth richest country in the world.\n\"The inequality has got worse. 150 billionaires in Britain, more people poor than ever.”\nThe audience applauded.\nHe added: “We’ve gone too far down the road of free market economics… we’ve created too much inequality.”\nWhen the leaders were asked about balancing the needs of security against human rights against in the wake of the London Bridge terror attack, Mr Corbyn said: \"What happened on the streets of London was utterly appalling and I very moved by what Jack Merritt\'s father said about what his son was trying to do.\nMr Johnson had threatened to lock up terrorists and throw away the key - even though victim Jack Merritt’s dad had pleaded for him not to politicise the atrocity.\nMr Corbyn slammed the Tory leader: “Yes of course you must imprison them. Yes of course you must try and rehabilitate them if you can.\n“But there has to be a monitoring process to achieve that. We need security on our streets. Security doesn’t come on the cheap.\n“And if you cut the number of police officers, cut the number of PCSOs, privatise the probation service, underfund all the other aspects of the criminal justice system, then we all pay a price with our own security.”\nMr Johnson claimed he was putting a huge amount of money into the criminal justice system - even though the Tories have cut expenditure on justice by 40%, and have announced no plans to increase expenditure except on prisons.\nBoris Johnson seemed to blame Islamophobia in the Conservative Party on the failure to get Brexit done.\nThe leaders were asked about the division and acrimony within political discourse by a member of the audience in Maidstone in Kent.\nThe Prime Minister replied: \"What we need to do is come together as a country.\n\"We\'ve been going at this too long - and this is my key point tonight - the scratchiness of politics, the bitterness, the acrimony is a function of our inability to get Brexit done.\"\nMr Johnson agreed to an independent investigation into Islamophobia in the Conservative Party while standing to be Tory leader before widening it out to include all forms of racism.",
            date_updated: 2019-12-07T09:53:43Z,
            date_published: 2019-12-06T23:21:21Z,
        },
    ),
)
